File: Betting app/BetTracker.swift
// BetTracker.swift
// Personal bet tracking — the user's own P&L on Pick1 calls.
//
// The app's core promise is "track everything"; this turns the public
// ledger into the USER's ledger. Tapping "Track bet" on a pick stores a
// row in `user_bets` (owner-only RLS); results inherit from the pick's
// win/loss grade, profit = stake × (odds − 1) on wins, −stake on losses.
// Stake is optional — a bet tracked without an amount still counts toward
// the personal record, just not the dollar P&L.

import Foundation
import Combine
import Supabase

struct UserBet: Codable, Identifiable, Hashable {
    let id: UUID
    let pickId: UUID
    let stake: Double?
    let oddsAtBet: Double?
    let createdAt: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case pickId = "pick_id"
        case stake
        case oddsAtBet = "odds_at_bet"
        case createdAt = "created_at"
    }
}

@MainActor
final class BetTracker: ObservableObject {
    static let shared = BetTracker()
    private init() {}

    /// pickId → bet. Drives instant "tracked" state on detail pages.
    @Published private(set) var bets: [UUID: UserBet] = [:]
    @Published private(set) var loaded = false

    private var client: SupabaseClient { SupabaseManager.client }

    func isTracked(_ pickId: UUID) -> Bool { bets[pickId] != nil }

    func load() async {
        guard client.auth.currentSession != nil else { return }
        do {
            let rows: [UserBet] = try await client.from("user_bets")
                .select("id, pick_id, stake, odds_at_bet, created_at")
                .execute().value
            bets = Dictionary(rows.map { ($0.pickId, $0) }, uniquingKeysWith: { a, _ in a })
            loaded = true
        } catch { /* non-fatal — tracker UI just shows untracked */ }
    }

    /// Track a pick, locking in today's odds. Upserts so re-tracking with a
    /// different stake just updates the row.
    func track(pick: Pick, stake: Double?) async {
        guard let uid = client.auth.currentSession?.user.id else { return }
        struct Row: Encodable {
            let user_id: String, pick_id: String
            let stake: Double?, odds_at_bet: Double?
        }
        let odds = pick.marketOdds ?? pick.impliedOddsForPayout
        do {
            try await client.from("user_bets")
                .upsert(Row(user_id: uid.uuidString, pick_id: pick.id.uuidString,
                            stake: stake, odds_at_bet: odds),
                        onConflict: "user_id,pick_id")
                .execute()
            await load()
            Analytics.track("bet_tracked", ["sport": pick.sport, "stake": stake ?? 0])
        } catch { }
    }

    func untrack(pickId: UUID) async {
        do {
            try await client.from("user_bets").delete()
                .eq("pick_id", value: pickId.uuidString).execute()
            bets[pickId] = nil
        } catch { }
    }

    // MARK: - P&L

    struct Summary {
        var tracked = 0
        var settled = 0
        var wins = 0
        var losses = 0
        var staked: Double = 0        // settled bets with a stake
        var profit: Double = 0        // net across settled staked bets
        var roiPct: Double? { staked > 0 ? profit / staked * 100 : nil }
    }

    /// Compute the personal record against the picks the bets refer to.
    func summary(picks: [Pick]) -> Summary {
        let byId = Dictionary(picks.map { ($0.id, $0) }, uniquingKeysWith: { a, _ in a })
        var s = Summary()
        for bet in bets.values {
            guard let pick = byId[bet.pickId] else { continue }
            s.tracked += 1
            guard !pick.isPending else { continue }
            s.settled += 1
            if pick.isWin { s.wins += 1 } else if pick.isLoss { s.losses += 1 }
            guard let stake = bet.stake, stake > 0 else { continue }
            let odds = bet.oddsAtBet ?? pick.marketOdds ?? 1.9
            s.staked += stake
            s.profit += pick.isWin ? stake * (odds - 1) : -stake
        }
        return s
    }
}

extension Pick {
    /// Fallback decimal odds when no market quote exists — derived from the
    /// model's own probability (used only for payout math, labeled as such).
    var impliedOddsForPayout: Double? {
        probability > 0 ? (100.0 / Double(probability)) : nil
    }
}

File: android/app/src/main/java/com/pick1/app/data/BetTracker.kt
package com.pick1.app.data

import com.pick1.app.data.model.Pick
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.postgrest.from
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Personal bet tracking — port of `BetTracker.swift`.
 *
 * Reads/writes the SAME `user_bets` table as iOS (owner-only RLS), so a
 * user's ledger follows them across platforms. Results inherit from the
 * pick's win/loss grade: profit = stake x (odds - 1) on wins, -stake on
 * losses. Stake is optional — a bet tracked without an amount still counts
 * toward the personal record, just not the dollar P&L.
 */
@Serializable
data class UserBet(
    val id: String,
    @SerialName("pick_id") val pickId: String,
    val stake: Double? = null,
    @SerialName("odds_at_bet") val oddsAtBet: Double? = null,
    @SerialName("created_at") val createdAt: String? = null,
)

data class BetSummary(
    val tracked: Int = 0,
    val settled: Int = 0,
    val wins: Int = 0,
    val losses: Int = 0,
    val staked: Double = 0.0,     // settled bets that carried a stake
    val profit: Double = 0.0,     // net across settled staked bets
) {
    val roiPct: Double? get() = if (staked > 0) profit / staked * 100 else null
    val pending: Int get() = tracked - settled
    val hitRate: Int? get() = if (settled > 0) ((wins.toDouble() / settled) * 100).toInt() else null
}

class BetRepository {

    suspend fun load(): Map<String, UserBet> =
        runCatching {
            Supabase.client.from("user_bets").select().decodeList<UserBet>()
                .associateBy { it.pickId }
        }.getOrDefault(emptyMap())

    suspend fun track(pick: Pick, stake: Double?): Boolean = runCatching {
        val uid = Supabase.client.auth.currentUserOrNull()?.id ?: return false
        Supabase.client.from("user_bets").insert(
            mapOf(
                "user_id" to uid,
                "pick_id" to pick.id,
                "stake" to stake?.toString(),
                "odds_at_bet" to pick.decimalOdds.toString(),
            )
        )
        true
    }.getOrDefault(false)

    /** Compute the personal record against the picks the bets refer to. */
    fun summary(bets: Collection<UserBet>, picks: List<Pick>): BetSummary {
        val byId = picks.associateBy { it.id }
        var tracked = 0; var settled = 0; var wins = 0; var losses = 0
        var staked = 0.0; var profit = 0.0
        for (bet in bets) {
            val pick = byId[bet.pickId] ?: continue
            tracked++
            if (pick.isPending) continue
            settled++
            if (pick.isWin) wins++ else if (pick.isLoss) losses++
            val stake = bet.stake ?: continue
            if (stake <= 0) continue
            val odds = bet.oddsAtBet ?: pick.marketOdds ?: 1.9
            staked += stake
            profit += if (pick.isWin) stake * (odds - 1) else -stake
        }
        return BetSummary(tracked, settled, wins, losses, staked, profit)
    }
}

// ── Calibration ──────────────────────────────────────────────────────

/**
 * A stated-confidence band vs the actual hit rate, straight from the public
 * `calibration_bands` view. Port of `CalibrationBand` in CalibrationView.swift.
 */
@Serializable
data class CalibrationBand(
    val band: String,
    val n: Int,
    val wins: Int,
    @SerialName("actual_pct") val actualPct: Double,
    @SerialName("avg_stated") val avgStated: Double,
)

class CalibrationRepository {
    suspend fun bands(): List<CalibrationBand> =
        runCatching {
            Supabase.client.from("calibration_bands").select()
                .decodeList<CalibrationBand>()
                .sortedByDescending { it.avgStated }   // highest confidence first
        }.getOrDefault(emptyList())

    /** The "within N points" honesty headline, weighted by sample size. */
    fun avgGapText(bands: List<CalibrationBand>): String? {
        if (bands.isEmpty()) return null
        val totalN = bands.sumOf { it.n }
        if (totalN <= 0) return null
        val weighted = bands.sumOf { kotlin.math.abs(it.avgStated - it.actualPct) * it.n } / totalN
        return weighted.toInt().toString()
    }
}
